### PR TITLE
Focus and hover improvements

### PR DESCRIPTION
**Pull request by myovchev**
## Summary

- fix missing required prop (probably here after the sync with main)
- add hover/focus Layout behavior according to the UX feedback (PRO-8326, PRO-8327)
- show resize/move indicators when appropriate (UX feedback, PRO-8329)

Ref  PRO-8326, PRO-8327, PRO-8329

### CODE REVIEW

File: modules/@apostrophecms/layout-widget/ui/apos/components/AposAreaLayoutEditor.vue [hunk 1]
```diff
@@ -35,6 +35,7 @@
       <template v-else>
         <AposAreaMenu
           :context-menu-options="contextMenuOptions"
+          :field-id="fieldId"
```
Comment (myovchev) [inline modules/@apostrophecms/layout-widget/ui/apos/components/AposAreaLayoutEditor.vue:38]: A new shiny bug (Vue dev warning about missing required prop). Most probably due to changes from `main`. It's good now.

File: modules/@apostrophecms/layout-widget/ui/apos/components/AposGridLayout.vue [hunk 2]
```diff
@@ -280,6 +280,7 @@ export default {
 /* stylelint-disable-next-line media-feature-name-allowed-list */
 @media (prefers-reduced-motion: no-preference) {
   /* TransitionGroup move/enter/leave animations for grid items */
+  // FIXME: casues issues on item resizing/re-rendering, works fine when moving
```
Comment (myovchev) [inline modules/@apostrophecms/layout-widget/ui/apos/components/AposGridLayout.vue:283]: There is very bad animation happening on resize and sometimes on widget rerendering (I think when there is a scroll and scrollTo operation). This has to be discussed with @stuartromanek before the final feature merge.

File: modules/@apostrophecms/layout-widget/ui/apos/components/AposAreaLayoutEditor.vue [hunk 3]
```diff
@@ -130,6 +131,10 @@ export default {
     };
   },
   computed: {
+    ...mapState(useWidgetStore, {
+      focusedWidgetState: state => state.focusedWidget,
+      hoveredWidgetState: state => state.hoveredWidget
+    }),
```
Comment (ValJed) [inline modules/@apostrophecms/layout-widget/ui/apos/components/AposAreaLayoutEditor.vue]: You should just get and use `focusedWidget` and `hoveredWidget`.
```js
    ...mapState(useWidgetStore, ['focusedWidget', 'hoveredWidget'])
```
This is already dynamic no need for new computed.

File: modules/@apostrophecms/layout-widget/ui/apos/components/AposAreaLayoutEditor.vue [hunk 4]
```diff
@@ -166,6 +174,21 @@ export default {
       return this.layoutModuleOptions.columnWidgetName;
     }
   },
+  watch: {
+    focusedWidgetState(widgetId) {
```
Comment (ValJed) [inline modules/@apostrophecms/layout-widget/ui/apos/components/AposAreaLayoutEditor.vue]: Could just watch `focusedWidget`.

File: modules/@apostrophecms/layout-widget/ui/apos/components/AposAreaLayoutEditor.vue [hunk 5]
```diff
@@ -166,6 +174,21 @@ export default {
...
+      if (this.layoutColumnWidgetIds.includes(widgetId)) {
+        this.clickOnGrid();
+      }
+    },
+    hoveredWidgetState(widgetId) {
```
Comment (ValJed) [inline modules/@apostrophecms/layout-widget/ui/apos/components/AposAreaLayoutEditor.vue]: Same for `hoveredWidget`

File: modules/@apostrophecms/layout-widget/ui/apos/components/AposGridManager.vue [hunk 6]
```diff
@@ -266,7 +339,11 @@ export default {
       },
       onMoveDebounced: throttle(this.onMove, 10),
       onResizeDebounced: throttle(this.onResize, 10),
-      lastPreviewKey: null
+      lastPreviewKey: null,
+      ghostHandleOffsets: {
+        west: null,
+        east: null
+      }
```
Comment (ValJed) [inline modules/@apostrophecms/layout-widget/ui/apos/components/AposGridManager.vue:346]: I would use `x` and `y` by convention.

File: modules/@apostrophecms/layout-widget/ui/apos/components/AposAreaLayoutEditor.vue (hunk 3 above)
Comment (myovchev) [reply modules/@apostrophecms/layout-widget/ui/apos/components/AposAreaLayoutEditor.vue]: We can't because of conflict in reactive names in the mixin (data). This is a safety change

File: modules/@apostrophecms/layout-widget/ui/apos/components/AposAreaLayoutEditor.vue (hunk 4 above)
Comment (myovchev) [reply modules/@apostrophecms/layout-widget/ui/apos/components/AposAreaLayoutEditor.vue]: We watch the store, it's namespaced due to the above reasons.

File: modules/@apostrophecms/layout-widget/ui/apos/components/AposAreaLayoutEditor.vue (hunk 5 above)
Comment (myovchev) [reply modules/@apostrophecms/layout-widget/ui/apos/components/AposAreaLayoutEditor.vue]: Same reason

File: modules/@apostrophecms/layout-widget/ui/apos/components/AposGridManager.vue (hunk 6 above)
Comment (myovchev) [reply modules/@apostrophecms/layout-widget/ui/apos/components/AposGridManager.vue:346]: West and east is by convention. It has nothing to do with x, y (horizontal, vertical)

File: modules/@apostrophecms/area/ui/apos/components/AposAreaWidget.vue [hunk 7]
```diff
@@ -291,11 +291,6 @@ export default {
       type: Boolean,
       default: false
     },
-    // Whether or not this widget should be focused
-    shouldFocus: {
-      type: Boolean,
-      default: true
-    },
```
Comment (myovchev) [inline modules/@apostrophecms/area/ui/apos/components/AposAreaWidget.vue:298]: Revert a stupid idea.

File: modules/@apostrophecms/layout-widget/ui/apos/components/AposAreaLayoutEditor.vue [hunk 8]
```diff
@@ -84,7 +85,6 @@
             :widget-focused="focusedWidget"
             :max-reached="maxReached"
             :rendering="rendering(widget)"
-            :should-focus="false"
```
Comment (myovchev) [inline modules/@apostrophecms/layout-widget/ui/apos/components/AposAreaLayoutEditor.vue:87]: Revert a stupid idea.

File: modules/@apostrophecms/area/ui/apos/components/AposAreaWidget.vue [hunk 9]
```diff
@@ -389,10 +385,6 @@ export default {
       return (this.widgetModuleOptions.widgetBreadcrumbOperations || []);
     },
     isFocused() {
-      if (!this.shouldFocus) {
-        return false;
-      }
-
```
Comment (myovchev) [inline modules/@apostrophecms/area/ui/apos/components/AposAreaWidget.vue:395]: Revert a stupid idea

Comment (stuartromanek) [reply modules/@apostrophecms/area/ui/apos/components/AposAreaWidget.vue:395]: no negative self-talk!!!!

**Review by ValJed — APPROVED** [review]

File: modules/@apostrophecms/area/ui/apos/lib/walk-widgets.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+  }
+
+  const stack = [ obj ];
+
+  while (stack.length > 0) {
+    const current = stack.pop();
+
+    if (!current || typeof current !== 'object') {
+      continue;
+    }
```
Comment (ValJed) [inline modules/@apostrophecms/area/ui/apos/lib/walk-widgets.js:33]: typeof [] returns object too but I guess no reason for this to happen?

File: modules/@apostrophecms/area/ui/apos/lib/walk-widgets.js [hunk 11]
```diff
@@ -0,0 +1,57 @@
+// Walk through an object and invoke the iterator function for every widget found
```
Comment (ValJed) [inline modules/@apostrophecms/area/ui/apos/lib/walk-widgets.js:1]: There is already a method `walk` in the area module. It takes a doc and an iterator.
Don't you think it could be used (it's a backend one). Or at least code could be shared between both?

File: modules/@apostrophecms/layout-widget/ui/apos/components/AposAreaLayoutEditor.vue [hunk 12]
```diff
@@ -166,6 +179,37 @@ export default {
...
   },
+  watch: {
+    // Steal the columns focus, set it on the layout widget instead.
+    // Additionally send "emphasized" state to the central store
+    // to keep the breadcrumb label visible, when children are focused.
+    async focusedWidget(widgetId) {
+      if (!this.parentOptions.widgetId) {
+        return;
+      }
+      await this.$nextTick();
```
Comment (ValJed) [inline modules/@apostrophecms/layout-widget/ui/apos/components/AposAreaLayoutEditor.vue:190]: Weird and annoying to need this but if only this works.
SetFocusedWidget should not be called to often, It don't really see how race condition could occur here

File: modules/@apostrophecms/layout-widget/ui/apos/components/AposAreaLayoutEditor.vue [hunk 13]
```diff
@@ -166,6 +179,37 @@ export default {
...
+        this.removeEmphasizedWidget(this.parentOptions.widgetId);
+      }
+    },
+    // Steal the columns hover, set it on the layout widget instead.
+    hoveredWidget(widgetId) {
+      if (
+        this.parentOptions.widgetId &&
+        this.layoutColumnWidgetIds.includes(widgetId)
+      ) {
+        this.setHoveredWidget(this.parentOptions.widgetId, null);
```
Comment (ValJed) [inline modules/@apostrophecms/layout-widget/ui/apos/components/AposAreaLayoutEditor.vue]:
```js
        this.setHoveredWidget(this.parentOptions.widgetId, this.areaId);
``` 
?

File: modules/@apostrophecms/layout-widget/ui/apos/components/AposGridManager.vue [hunk 14]
```diff
@@ -671,6 +763,67 @@ export default {
       } else {
         // todo should never be missing!
       }
+    },
+    resetGhostHandleOffsets() {
+      this.ghostHandleOffsets.west = null;
+      this.ghostHandleOffsets.east = null;
+    },
+    updateGhostResizeGrip(side, pointerY) {
+      const key = side === 'west' ? 'west' : 'east';
```
Comment (ValJed) [inline modules/@apostrophecms/layout-widget/ui/apos/components/AposGridManager.vue]: So key === side ? why not directly using side param?

File: modules/@apostrophecms/layout-widget/ui/apos/components/AposGridManager.vue [hunk 15]
```diff
@@ -671,6 +763,67 @@ export default {
...
+      if (typeof pointerY !== 'number') {
```
Comment (ValJed) [inline modules/@apostrophecms/layout-widget/ui/apos/components/AposGridManager.vue:772]: Mixing east / west and X / Y ?

File: modules/@apostrophecms/layout-widget/ui/apos/components/AposGridManager.vue [hunk 16]
```diff
@@ -671,6 +763,67 @@ export default {
...
+        return;
+      }
+      const height = this.ghostData.height ?? 0;
+      if (!height && !this.ghostData.element) {
+        return;
+      }
```
Comment (ValJed) [inline modules/@apostrophecms/layout-widget/ui/apos/components/AposGridManager.vue:779]: Shouldn't it be || instead of && ?

File: modules/@apostrophecms/layout-widget/ui/apos/components/AposGridManager.vue [hunk 17]
```diff
@@ -671,6 +763,67 @@ export default {
...
+      const effectiveHeight = height || this.ghostData.element
+        .getBoundingClientRect()
+        ?.height || RESIZE_ICON_HEIGHT;
```
Comment (ValJed) [inline modules/@apostrophecms/layout-widget/ui/apos/components/AposGridManager.vue]: What about:
```js
const  effectiveHeight = height
  ? height
  : this.ghostData.element?.getBoundingClientRect();
 if (!height) {return}
```

File: modules/@apostrophecms/area/ui/apos/lib/walk-widgets.js (hunk 10 above)
Comment (myovchev) [reply modules/@apostrophecms/area/ui/apos/lib/walk-widgets.js:33]: Array is a legit object in this case. We want every "object" but null.

File: modules/@apostrophecms/area/ui/apos/lib/walk-widgets.js (hunk 11 above)
Comment (myovchev) [reply modules/@apostrophecms/area/ui/apos/lib/walk-widgets.js:1]: The one I implemented is fast and lightweight. The existing backend walker is more heavy (has more features that we don't care about). No need of sharing.

File: modules/@apostrophecms/layout-widget/ui/apos/components/AposAreaLayoutEditor.vue (hunk 12 above)
Comment (myovchev) [reply modules/@apostrophecms/layout-widget/ui/apos/components/AposAreaLayoutEditor.vue:190]: You have "bulk" reactive racing operations: when focus is reported, the widget state is available on the next tick.

File: modules/@apostrophecms/layout-widget/ui/apos/components/AposAreaLayoutEditor.vue (hunk 13 above)
Comment (myovchev) [reply modules/@apostrophecms/layout-widget/ui/apos/components/AposAreaLayoutEditor.vue]: I'm not sure what's the reason to provide the area ID, we only want to steal a "hover" from the column widget. I'll add it but I'm curious what happens behind the scenes?

File: modules/@apostrophecms/layout-widget/ui/apos/components/AposGridManager.vue (hunk 14 above)
Comment (myovchev) [reply modules/@apostrophecms/layout-widget/ui/apos/components/AposGridManager.vue]: Yes, a typo, leftover after internal refactoring. Fixed

File: modules/@apostrophecms/layout-widget/ui/apos/components/AposGridManager.vue (hunk 15 above)
Comment (myovchev) [reply modules/@apostrophecms/layout-widget/ui/apos/components/AposGridManager.vue:772]: Yes, horizontal side (east, west by convention) and vertical position (x, y) are different concepts, both participating in this computation. 
We use x,y only for positions. For directions we use e.g. `s` (South) or `se` or `sw` etc.

File: modules/@apostrophecms/layout-widget/ui/apos/components/AposGridManager.vue (hunk 17 above)
Comment (myovchev) [reply modules/@apostrophecms/layout-widget/ui/apos/components/AposGridManager.vue]: Improved further

File: modules/@apostrophecms/area/ui/apos/lib/walk-widgets.js (hunk 10 above)
Comment (myovchev) [reply modules/@apostrophecms/area/ui/apos/lib/walk-widgets.js:33]: All edge cases are covered by unit tests.
